alf/environments/wrappers.py

```diff
@@ -27,44 +30,161 @@
 
-    def __init__(self, env, stack_size=4, channel_order='channels_last'):
+    def __init__(
+            self,
+            env,
+            stack_size=4,
+            channel_order='channels_last',
+            fields_to_stack=None,
+    ):
         """Create a FrameStack object.
 
         Args:
+            env (gym.Space): gym environment.
             stack_size (int):
             channel_order (str): one of `channels_last` or `channels_first`.
                 The ordering of the dimensions in the images.
                 `channels_last` corresponds to images with shape
                 `(height, width, channels)` (Atari's default) while
                 `channels_first` corresponds to images with shape
                 `(channels, height, width)`.
+            fields_to_stack (list of str): optional paths to the fields of the
+                Dict env.observation_space.  If specified, only use the
+                spaces corresponding to the keys in FrameStack.  If is None,
+                stack all recognized spaces.
+                When input space is not a Dict, we just try to stack if we can.
+                Fields_to_stack doesn't apply anymore.
         """
         super().__init__(env)
-        self._frames = collections.deque(maxlen=stack_size)
+        self._frames = deque(maxlen=stack_size)
         self._channel_order = channel_order
         self._stack_size = stack_size
-
-        space = self.env.observation_space
-        assert channel_order in ['channels_last', 'channels_first']
-
-        if channel_order == 'channels_last':
-            shape = list(space.shape[0:-1]) + [stack_size * space.shape[-1]]
+        raw_space = self.env.observation_space
+
+        if not fields_to_stack and isinstance(raw_space, gym.spaces.Dict):
+            fields_to_stack = raw_space.spaces.keys()
+
+        self._fields_to_stack = fields_to_stack
+
+        def _stack_space(sp):
+            if isinstance(sp, gym.spaces.Box):
+                # Shape of stacked_space is determined by low.shape
+                low = np.concatenate([sp.low] * stack_size)
+                high = np.concatenate([sp.high] * stack_size)
+                assert channel_order in ['channels_last', 'channels_first']
+                if channel_order == 'channels_last':
+                    low = np.transpose(
+                        np.concatenate([np.transpose(sp.low)] * stack_size))
+                    high = np.transpose(
+                        np.concatenate([np.transpose(sp.high)] * stack_size))
+                stacked_space = gym.spaces.Box(
+                    low=np.array(low), high=np.array(high), dtype=sp.dtype)
+
+            elif isinstance(sp, gym.spaces.MultiDiscrete):
+                nvec = [sp.nvec] * stack_size
+                stacked_space = gym.spaces.MultiDiscrete(nvec)
+
+            else:
+                stacked_space = sp
+            return stacked_space
+
+        def _traverse(d, fields_to_stack=None, prefix=""):
```

Comment:
It seems that this function will change the original space of wrapped the environment. We should make a copy so that the space of the wrapped environment will not be changed.

Reply:
Good catch!  Done, and this will break the new test added.